Documentation/architecture.rst

```diff
@@ -0,0 +1,155 @@
+
+The following is a list of the hooks used by a Cilium and a brief
+description. For a more thorough documentation on specifics of each
+hook see :ref:`bpf_guide`.
+
+* **XDP:** The XDP BPF hook is at the earliest point possible in the networking driver
+  and triggers a run of the BPF program upon packet reception. This
+  achieves the best possible packet processing performance since the
+  program runs directly on the packet data before any other processing
+  can happen. This hook is ideal for running filtering programs that
+  drop malicious or unexpected traffic and other common DDOS protection
+  mechanisms.
+
+* **Traffic Control Ingress:** BPF programs attached to the traffic control (tc)
+  ingress hook are attached to a networking interface, same as XDP, but
+  will run after the networking stack has done initial processing
+  of the packet. The hook is run before the L3 layer of the stack but has access
+  to most of the metadata associated with a packet. This is ideal for
+  doing local node processing, such as applying L3/L4 endpoint policy
+  and redirecting traffic to endpoints. For networking facing devices the
+  tc ingress hook can be coupled with above XDP hook. When this is done  it
+  is reasonable to assume that the majority of the traffic at this
+  point is legitimate and destined for the host. Containers use a virtual device
+  called a veth pair which acts as a virtual wire connecting the container to
+  the host. By attaching to the TC ingress hook of the host side of this veth pair
+  Cilium can monitor and enforce policy on all traffic exiting a container.
+  By attaching a BPF program to the veth pair from all containers and routing
+  all network traffic to a host side virtual device with another BPF program attached to
+  the tc ingress hook as well Cilium can monitor and enforce policy on all traffic
+  entering/exiting the node.
+
+* **Socket operations:** The socket operations hook is attached to a specific
+  cgroup and runs on TCP events. Cilium attaches a BPF Socket operations
+  program to the root cgroup and uses this to monitor for TCP state transitions,
+  specifically for ESTABLISHED state transitions. When
+  a socket transitions into ESTABLISH state if the TCP socket has a node
+  local peer (possibly a local proxy) a Socket send/recv program is attached.
+
+* **Socket send/recv:** The socket send/recv hook runs on every send operation
+  performed by a TCP socket. At this point the hook can inspect the message
+  and either drop the message, send the message to the TCP layer, or redirect
+  the message to another socket. Cilium uses this to accelerate the datapath redirects
+  as described below.
+
+Combining the above hooks with a virtual interfaces (cilium_host, cilium_net),
+an optional overlay interface (cilium_vxlan) and a userspace proxy (Envoy) Cilium
+creates the following networking objects.
+
+* **Prefilter:** The prefilter object runs an XDP program for best performance and
```

Comment:
the "for best performance" bit here seems a little early to describe since the prefilter isn't even described yet ;-) Might make sense to push that fragment a little later, even just reworded at the end of the sentence.

Reply:
moving fragment to the end works.